src/installer.ts

```diff
@@ -1,54 +1,103 @@
 import * as semver from 'semver';
-
-let osPlat: string = os.platform();
-let osArch: string = translateArchToDistUrl(os.arch());
+import {Url} from 'url';
+import fs = require('fs');
 
 //
 // Node versions interface
 // see https://nodejs.org/dist/index.json
 //
-interface INodeVersion {
+export interface INodeVersion {
   version: string;
   files: string[];
 }
 
-export async function getNode(versionSpec: string) {
+interface INodeVersionInfo {
+  downloadUrl: string;
+  token: string | null;
+  resolvedVersion: string;
+  fileName: string;
+}
+
+export async function getNode(
+  versionSpec: string,
+  stable: boolean,
+  token: string
+) {
+  let osPlat: string = os.platform();
+  let osArch: string = translateArchToDistUrl(os.arch());
+
   // check cache
+  let info: INodeVersionInfo | null = null;
   let toolPath: string;
   toolPath = tc.find('node', versionSpec);
 
   // If not found in cache, download
-  if (!toolPath) {
-    let version: string;
-    const c = semver.clean(versionSpec) || '';
-    // If explicit version
-    if (semver.valid(c) != null) {
-      // version to download
-      version = versionSpec;
-    } else {
-      // query nodejs.org for a matching version
-      version = await queryLatestMatch(versionSpec);
-      if (!version) {
-        throw new Error(
-          `Unable to find Node version '${versionSpec}' for platform ${osPlat} and architecture ${osArch}.`
-        );
+  if (toolPath) {
+    console.log(`Found in cache @ ${toolPath}`);
+  } else {
+    console.log(`Attempting to download ${versionSpec}...`);
+    let info = await getInfoFromManifest(versionSpec, stable, token);
+    if (!info) {
+      console.log(
+        'Not found in manifest.  Falling back to download directly from Node'
+      );
+      info = await getInfoFromDist(versionSpec);
+    }
+
+    if (!info) {
+      throw new Error(
+        `Unable to find Node version '${versionSpec}' for platform ${osPlat} and architecture ${osArch}.`
+      );
+    }
+
+    console.log(`Acquiring ${info.resolvedVersion} from ${info.downloadUrl}`);
+
+    let downloadPath = '';
+    try {
+      downloadPath = await tc.downloadTool(info.downloadUrl, undefined, token);
+    } catch (err) {
+      if (err instanceof tc.HTTPError && err.httpStatusCode == 404) {
+        await acquireNodeFromFallbackLocation(info.resolvedVersion);
+        return;
       }
 
-      // check cache
-      toolPath = tc.find('node', version);
+      throw err;
     }
 
-    if (!toolPath) {
-      // download, extract, cache
-      toolPath = await acquireNode(version);
+    //
+    // Extract
+    //
+    console.log('Extracting ...');
+    let extPath: string;
+    if (osPlat == 'win32') {
+      let _7zPath = path.join(__dirname, '..', 'externals', '7zr.exe');
+      extPath = await tc.extract7z(downloadPath, undefined, _7zPath);
+      // 7z extracts to folder matching file name
+      let nestedPath = path.join(extPath, path.basename(info.fileName, '.7z'));
```

Comment:
node dist creates a 7z with a nested folder matching the filename in the unzipped structure.  The manifest extracts without that.  That's a good thing because later I'm considering extracting directly into the cache instead of unzip and copy.  Copy on windows consumes the vast majority of the time.

Reply:
It looks like unzip + copy may be related to the extra folder after unzipping.

Move is unsafe on Windows because defender commonly still has a lock reading the newly created files on disk. Copy is safe.